Prepare VolumeHost for running mount tools in containers

jsafrane opened the pull request:
This is the first part of implementation of https://github.com/kubernetes/features/issues/278 - running mount utilities in containers.

It updates `VolumeHost` interface:

*  `GetMounter()` now requires volume plugin name, as it is going to return different mounter to different volume plugings, because mount utilities for these plugins can be on different places.
* New `GetExec()` method that should volume plugins use to execute any utilities. This new `Exec` interface will execute them on proper place.
* `SafeFormatAndMount` is updated to the new `Exec` interface.

This is just a preparation, `GetExec` right now leads to simple `os.Exec` and mount utilities are executed on the same place as before. Also, the volume plugins will be updated in subsequent PRs (split into separate PRs, some plugins required lot of changes).

```release-note
NONE
```

@kubernetes/sig-storage-pr-reviews 
@rootfs @gnufied

jsafrane commented:
@dchen1107 @derekwaynecarr, can I ask you for a quick approval of kubelet changes? They're quite minimal, the real PR will come :-)

gnufied commented on pkg/util/mount/exec.go:
```diff
@@ -0,0 +1,50 @@
+
+var _ Exec = &osExec{}
+
+func (e *osExec) Run(cmd string, args []string) ([]byte, error) {
```
Should `Run` second argument be variadic to mimic the interface of `exe.Command` ?

jsafrane replied:
I mimicked `Kubelet.RunInContainer()`, but `exec.Command` is used in volume plugins, it would be probably easier to mimic `exec`... Update will follow shortly.

jsafrane replied:
fixed

rootfs commented:
@jsafrane do you have a preview of how `GetMounter()` passes mounters to different plugins? Does nsenter need a new mounter as well?

jsafrane commented:
@rootfs, here is the big picture with draft changes to kubelet and everything: https://github.com/jsafrane/kubernetes/commits/mount-container2

Kubelet chooses the right mounter here: https://github.com/jsafrane/kubernetes/blob/mount-container2/pkg/kubelet/volume_host.go#L119
And ExecMounter is here: https://github.com/jsafrane/kubernetes/blob/mount-container2/pkg/util/mount/exec_mount.go

Both are pretty straightforward. Note that the branch is now obsolete, there will be e.g. no cmdline option to configure namespace with the mount pods in the first experimental version.

rootfs commented:
@jsafrane It looks there is just mounter for flex plugin. Should all flex drivers share the same mount pod?

jsafrane commented:
@rootfs flex is not part of alpha implementation, it will exec things on host. In future, every flex can have its own pod.

rootfs commented:
@jsafrane sounds good
/assign
/lgtm

gnufied commented:
lgtm

jsafrane commented:
@k8s-bot pull-kubernetes-cross test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

rootfs commented on pkg/util/mount/mount_linux.go:
```diff
@@ -396,9 +394,8 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
-	cmd := mounter.Runner.Command("lsblk", args...)
 	glog.V(4).Infof("Attempting to determine if disk %q is formatted using lsblk with args: (%v)", disk, args)
-	dataOut, err := cmd.CombinedOutput()
+	dataOut, err := mounter.Exec.Run("lsblk", args...)
```
`findmnt` too (both in util-linux package)?

jsafrane replied:
I'm not sure what you mean by this comment. `findmnt` is not used by `mount_linux.go` and it has no use when finding a fs on a block device.

jsafrane commented:
Rebased

jsafrane commented:
rebased

jsafrane commented:
@jingxu97 @smarterclayton please approve

jingxu97 commented:
/approve

jingxu97 commented on pkg/kubelet/volume_host.go:
```diff
@@ -117,7 +117,7 @@ func (kvh *kubeletVolumeHost) GetCloudProvider() cloudprovider.Interface {
 }
 
-func (kvh *kubeletVolumeHost) GetMounter() mount.Interface {
+func (kvh *kubeletVolumeHost) GetMounter(pluginName string) mount.Interface {
```
You plan to use the pluginName in later PR?

jsafrane replied:
Yes, see https://github.com/kubernetes/community/pull/589

jsafrane committed:
Add Exec interface to VolumeHost

This exec should be used by volume plugins to execute mount utilities.
It will eventually execute things in mount containers.

jsafrane committed:
Add pluginName to VolumeHost.GetMouter

Different plugins can get different mounter, depending where the mount
utilities are.

jsafrane committed:
SafeFormatAndMount should use volume.Exec provided by VolumeHost

We need to execute mkfs / fsck where the utilities are.

jsafrane commented:
/assign @derekwaynecarr @vishh  @tallclair 
Please approve kubelet changes, the're quite minimal.

jsafrane commented:
/assign @Random-Liu @dchen1107 @derekwaynecarr @tallclair @vishh @yujuhong
Please approve kubelet changes, the're quite minimal.

derekwaynecarr commented:
the kubelet code changes look fine.

/approve

jsafrane commented:
@rootfs @jingxu97 please re-lgtm.

rootfs commented:
/lgtm
